File: src/main/rs/emulate/shared/util/crypto/Xtea.java
package rs.emulate.shared.util.crypto;

import rs.emulate.shared.util.DataBuffer;

/**
 * An implementation of the XTEA block cipher.
 *
 * @author Graham
 */
public final class Xtea {

	/**
	 * The golden ratio.
	 */
	public static final int GOLDEN_RATIO = 0x9E3779B9;

	/**
	 * The number of rounds.
	 */
	public static final int ROUNDS = 32;

	/**
	 * Deciphers the data in the specified {@link DataBuffer}.
	 * 
	 * @param buffer The buffer.
	 * @param start The start value.
	 * @param end The end value.
	 * @param key The key.
	 * @throws IllegalArgumentException if the key is not exactly 4 elements long.
	 */
	public static void decipher(DataBuffer buffer, int start, int end, int[] key) {
		if (key.length != 4) {
			throw new IllegalArgumentException("Key length must be four.");
		}

		int numQuads = (end - start) / 8;
		for (int i = 0; i < numQuads; i++) {
			int sum = GOLDEN_RATIO * ROUNDS;
			int v0 = buffer.getInt(start + i * 8);
			int v1 = buffer.getInt(start + i * 8 + 4);
			for (int j = 0; j < ROUNDS; j++) {
				v1 -= (v0 << 4 ^ v0 >>> 5) + v0 ^ sum + key[sum >>> 11 & 3];
				sum -= GOLDEN_RATIO;
				v0 -= (v1 << 4 ^ v1 >>> 5) + v1 ^ sum + key[sum & 3];
			}
			buffer.putInt(start + i * 8, v0);
			buffer.putInt(start + i * 8 + 4, v1);
		}
	}

	/**
	 * Enciphers the specified {@link DataBuffer} with the given key.
	 *
	 * @param buffer The buffer.
	 * @param start The start value.
	 * @param end The end value.
	 * @param key The key.
	 * @throws IllegalArgumentException if the key is not exactly 4 elements long.
	 */
	public static void encipher(DataBuffer buffer, int start, int end, int[] key) {
		if (key.length != 4) {
			throw new IllegalArgumentException("Key length must be four.");
		}

		int numQuads = (end - start) / 8;
		for (int i = 0; i < numQuads; i++) {
			int sum = 0;
			int v0 = buffer.getInt(start + i * 8);
			int v1 = buffer.getInt(start + i * 8 + 4);
			for (int j = 0; j < ROUNDS; j++) {
				v0 += (v1 << 4 ^ v1 >>> 5) + v1 ^ sum + key[sum & 3];
				sum += GOLDEN_RATIO;
				v1 += (v0 << 4 ^ v0 >>> 5) + v0 ^ sum + key[sum >>> 11 & 3];
			}
			buffer.putInt(start + i * 8, v0);
			buffer.putInt(start + i * 8 + 4, v1);
		}
	}

	/**
	 * Default private constructor to prevent instantiation.
	 */
	private Xtea() {

	}

}

File: util/src/main/kotlin/rs/emulate/shared/util/crypto/Xtea.kt
package rs.emulate.shared.util.crypto

import java.nio.ByteBuffer

/**
 * An implementation of the XTEA block cipher.
 */
object Xtea {

    /**
     * The golden ratio.
     */
    private const val GOLDEN_RATIO = 0x9E3779B9.toInt()

    /**
     * The number of rounds.
     */
    private const val ROUNDS = 32

    /**
     * Deciphers the data in the specified [ByteBuffer].
     * @throws IllegalArgumentException if the key is not exactly 4 elements long.
     */
    fun decipher(buffer: ByteBuffer, start: Int, end: Int, key: IntArray) {
        require(key.size == 4) { "Key length must be four." }

        val quads = (end - start) / 8
        for (index in 0 until quads) {
            var sum = GOLDEN_RATIO * ROUNDS
            var v0 = buffer.getInt(start + index * 8)
            var v1 = buffer.getInt(start + index * 8 + 4)

            for (j in 0 until ROUNDS) {
                v1 -= (v0 shl 4 xor v0.ushr(5)) + v0 xor sum + key[sum ushr 11 and 3]
                sum -= GOLDEN_RATIO
                v0 -= (v1 shl 4 xor v1.ushr(5)) + v1 xor sum + key[sum and 3]
            }

            buffer.putInt(start + index * 8, v0)
            buffer.putInt(start + index * 8 + 4, v1)
        }
    }

    /**
     * Enciphers the specified [ByteBuffer] with the given key.
     * @throws IllegalArgumentException if the key is not exactly 4 elements long.
     */
    fun encipher(buffer: ByteBuffer, start: Int, end: Int, key: IntArray) {
        require(key.size == 4) { "Key length must be four." }

        val quads = (end - start) / 8
        for (index in 0 until quads) {
            var sum = 0
            var v0 = buffer.getInt(start + index * 8)
            var v1 = buffer.getInt(start + index * 8 + 4)

            for (j in 0 until ROUNDS) {
                v0 += (v1 shl 4 xor v1.ushr(5)) + v1 xor sum + key[sum and 3]
                sum += GOLDEN_RATIO
                v1 += (v0 shl 4 xor v0.ushr(5)) + v0 xor sum + key[sum ushr 11 and 3]
            }

            buffer.putInt(start + index * 8, v0)
            buffer.putInt(start + index * 8 + 4, v1)
        }
    }

}
